docs/source/notes/autograd.rst

```diff
@@ -3,119 +3,394 @@
-By tracing this graph from roots to leaves, you can automatically
-compute the gradients using the chain rule.
-
-Internally, autograd represents this graph as a graph of
-:class:`Function` objects (really expressions), which can be
-:meth:`~torch.autograd.Function.apply` ed to compute the result of
-evaluating the graph.  When computing the forwards pass, autograd
-simultaneously performs the requested computations and builds up a graph
-representing the function that computes the gradient (the ``.grad_fn``
-attribute of each :class:`torch.Tensor` is an entry point into this graph).
-When the forwards pass is completed, we evaluate this graph in the
-backwards pass to compute the gradients.
-
-An important thing to note is that the graph is recreated from scratch at every
-iteration, and this is exactly what allows for using arbitrary Python control
-flow statements, that can change the overall shape and size of the graph at
-every iteration. You don't have to encode all possible paths before you
-launch the training - what you run is what you differentiate.
-
-In-place operations with autograd
+    def f(w):
+        x = op_1(w)
+        y = op_2(x)
+        z = op_3(y)
+        return z
+
+    w.requires_grad_()
+    f(w).backward(v)
+
+Will change the computation to: :math:`v^T J_f = v^T J_{op_3} J_{op_2} J_{op_1} = v^T \dfrac{\partial{z}}{\partial{y}} \dfrac{\partial{y}}{\partial{x}} \dfrac{\partial{x}}{\partial{w}} = v^T \dfrac{\partial{z}}{\partial{w}}`.
+
+Handling inplace operations
+"""""""""""""""""""""""""""
+
+The only special case that needs to be considered here are inplace operations.
+The main reason to use inplace operations is to preserve side effects and prevent additional memory requirements.
+The first point means that if a tensor is referenced in different places, the inplace operation should modify all these references.
+The second point means that no extra memory allocation should be done under the hood otherwise it would defeat the purpose of inplace operations.
+
+To be able to fulfill these two needs, we make inplace operations actually be inplace during the forward pass giving exactely the same behavior
+as any other non-autograd library that provides inplace operations.
+To be able to compute the gradients, we consider that there are different versions of the tensor that existed and that the user tensor always points to the latest version.
+This allows us to write the automatic differentiation rules as before.
+
+See the two functions below for examples where we use subscripts to specify each version of the Tensor.:
+
+.. code::
+
+    def f(x):
+        z = op_1(x)
+        # z points to z_0
+        z.op_2_()
+        # z points to z_1
+        return z
+
+    def g(x):
+        z = op_1(x)
+        # z points to z_0
+        y = z.select(0, 0)
+        # y is a view into z
+        # y points to y_0 = z_0.select(0, 0)
+        z.op_2_()
+        # both z and y now points to y_1 = z_1.select(0, 0)
```

Comment:
I like the example. Can you confirm why z now points to z.select(0, 0)?

Reply:
You mean "why y now points to z.select(0,0)" ?
Because it points to the same memory space as the latest version of z. So if one change, the other has to change as well.
Do you want to add that in the comment?

Reply:
I understand that "y now points to z_1.select(0, 0)". The comment on this line says that "both z and y now points to y_1 = z_1.select(0, 0)". And so I'm wondering if this indeed meant that "z now points z_1.select(0,0)". I believe you meant 
```
z.op_2_()
# z now points to z_1 = op_2(z_0)
# y now points to y_1 = z_1.select(0, 0)
```
Is that so? :)

Reply:
Yes ! I'll split that in two lines !